toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -907,33 +905,31 @@ auto TokenizedBuffer::SourceBufferLocationTranslator::GetLocation(
   int line_number = line_it - buffer_->line_infos_.begin();
   int column_number = offset - line_it->start;
 
-  llvm::StringRef line;
-
-  // We might still be lexing the last line. If so, check to see if there are
-  // any newline characters between the position we've finished lexing up to
-  // and the given location.
-  if (incomplete_line_info && column_number > *last_line_lexed_to_column_) {
-    column_number = *last_line_lexed_to_column_;
-    int64_t start = line_it->start;
-    for (int64_t i = line_it->start + *last_line_lexed_to_column_; i != offset;
-         ++i) {
-      if (buffer_->source_->text()[i] == '\n') {
-        start = i;
-        ++line_number;
-        column_number = 0;
-      } else {
-        ++column_number;
+  // Start by grabbing the buffer. If the line isn't fully lexed, the length
+  // will be npos and the buffer will be grabbed from the start; we'll then
+  // adjust the length.
+  llvm::StringRef line =
+      buffer_->source_->text().substr(line_it->start, line_it->length);
+  if (line_it->length == static_cast<int32_t>(llvm::StringRef::npos)) {
+    // The location of the diagnostic hasn't been lexed yet. Check to see if
+    // there are any newline characters between the position we've finished
+    // lexing up to and the given location.
+    if (column_number > *last_line_lexed_to_column_) {
+      auto start_newline_pos =
+          line.take_front(*last_line_lexed_to_column_).rfind('\n');
+      if (start_newline_pos != llvm::StringRef::npos) {
```

Comment:
Can this happen? It looks like this would mean there is a newline in the first `*last_line_lexed_to_column_` bytes of the last line, which seems like it should be impossible (if there's a newline there, we should have a later last line).

Reply:
I'd really thought there was something here, but I see my logic was off already. I've switched to a CHECK so that if this happens we can fix it.

Reply:
There can be newlines in `[*last_line_lexed_to_column_, column_number)`, but there can't be newlines in `[0, *last_line_lexed_to_column_)`.